pkg/action/executor/move_util.go

```diff
@@ -36,63 +40,141 @@ func copyPodInfo(oldPod, newPod *api.Pod) {
+	timeout := time.Duration(retryNum+1) * interval
+	err := goutil.RetrySimple(retryNum, timeout, interval, func() (bool, error) {
+		return doCheckPodNode(client, namespace, name, nodeName)
+	})
+	return err
+}
+
+// Move pod to node nodeName in three steps:
+//  step1: create a clone pod of the original pod (without labels)
+//  step2: delete the original pod;
+//  step3: add the labels to the cloned pod;
+func movePod(client *kclient.Clientset, tpod *api.Pod, nodeName string, retryNum int) (*api.Pod, error) {
+	podClient := client.CoreV1().Pods(tpod.Namespace)
 	if podClient == nil {
-		err := fmt.Errorf("cannot get Pod client for nameSpace:%v", pod.Namespace)
+		err := fmt.Errorf("cannot get Pod client for nameSpace:%v", tpod.Namespace)
 		glog.Error(err)
 		return nil, err
 	}
 
-	//1. copy the original pod
-	id := fmt.Sprintf("%v/%v", pod.Namespace, pod.Name)
-	glog.V(2).Infof("move-pod: begin to move %v from %v to %v",
-		id, pod.Spec.NodeName, nodeName)
+	// 0. get the latest Pod
+	var err error
+	var pod *api.Pod
+	if pod, err = podClient.Get(tpod.Name, metav1.GetOptions{}); err != nil {
+		glog.Errorf("Failed to get latest pod %v/%v: %v", tpod.Namespace, tpod.Name, err)
+		return nil, err
+	}
+	//NOTE: do deep-copy if the original pod may be modified outside this function
+	labels := pod.Labels
 
-	npod := &api.Pod{}
-	copyPodInfo(pod, npod)
-	npod.Spec.NodeName = nodeName
+	//1. create a clone pod--podC of the original pod--podA
+	npod, err := createClonePod(client, pod, nodeName)
+	if err != nil {
+		glog.Errorf("Move pod failed: failed to create a clone pod: %v", err)
+		return nil, err
+	}
+
+	//delete the clone pod if this action fails
+	flag := false
+	defer func() {
+		if !flag {
+			glog.Errorf("Move pod failed, begin to delete cloned pod: %v/%v", npod.Namespace, npod.Name)
+			delOpt := &metav1.DeleteOptions{}
+			podClient.Delete(npod.Name, delOpt)
+		}
+	}()
 
-	//2. kill original pod
-	grace := calcGracePeriod(pod)
-	delOption := &metav1.DeleteOptions{GracePeriodSeconds: &grace}
-	err := podClient.Delete(pod.Name, delOption)
+	//1.2 wait until podC gets ready
+	err = waitForReady(client, npod.Namespace, npod.Name, nodeName, retryNum)
 	if err != nil {
-		err = fmt.Errorf("move-failed: failed to delete original pod-%v: %v",
-			id, err)
-		glog.Error(err)
+		glog.Errorf("Wait for cloned Pod ready timeout: %v", err)
 		return nil, err
 	}
 
-	//3. create (and bind) the new Pod
-	time.Sleep(time.Duration(grace)*time.Second + defaultMoreGrace) //wait for the previous pod to be cleaned up.
-	interval := defaultPodCreateSleep
-	timeout := time.Duration(retryNum+1) * interval
-	err = util.RetryDuring(retryNum, timeout, interval, func() error {
-		_, inerr := podClient.Create(npod)
-		return inerr
-	})
+	//2. delete the original pod--podA
+	delOpt := &metav1.DeleteOptions{}
+	if err := podClient.Delete(pod.Name, delOpt); err != nil {
+		glog.Warningf("Move pod warning: failed to delete original pod: %v", err)
```

Comment:
Shouldn't it be an error message since failure deleting the pod causes some problem?